src/vcpkg/base/system.process.cpp

```diff
@@ -677,17 +677,35 @@ namespace vcpkg
             cmd_line,
             wd,
             [&](StringView sv) {
-                auto prev_size = buf.size();
-                Strings::append(buf, sv);
-
-                auto it = std::find(buf.begin() + prev_size, buf.end(), '\n');
-                while (it != buf.end())
+                const auto e = sv.end();
+                auto cur_begin = sv.begin();
+                auto it = std::find(cur_begin, e, '\n');
+                if (it != e)
                 {
-                    std::string s(buf.begin(), it);
-                    per_line_cb(s);
-                    buf.erase(buf.begin(), it + 1);
-                    it = std::find(buf.begin(), buf.end(), '\n');
+                    {
+                        const auto next_begin = it + 1;
+                        buf.append(cur_begin, it);
+                        // omit trailing \r
+                        if (!buf.empty() && buf.back() == '\r') buf.pop_back();
+                        per_line_cb(buf);
+                        buf.clear();
+                        cur_begin = next_begin;
```

Comment:
This is a little complex for no reason; we can clear it up quite a bit. Additionally, I would like a comment here, because I was pretty confused my first time reading it:

```suggestion
                        // include the start of this line from the last callback
                        buf.append(current_start, it);
                        // omit trailing \r
                        if (!buf.empty() && buf.back() == '\r') buf.pop_back();
                        per_line_cb(buf);
                        buf.clear();
                        current_start = it + 1;
```

the `next_begin` variable feels unnecessary, so I have removed it in my suggestion. This also means we could remove the scope.